File: daily/Backtracking/90-Subsets-II.go
package daily

import "sort"

func subsetsWithDup(nums []int) [][]int {
	var results [][]int
	var path []int
	sort.Ints(nums)

	var backtrack func(start int)
	backtrack = func(start int) {
		subset := make([]int, len(path))
		copy(subset, path)
		results = append(results, subset)

		for i := start; i < len(nums); i++ {
			if i > start && nums[i-1] == nums[i] {
				continue
			}
			path = append(path, nums[i])
			backtrack(i + 1)
			path = path[:len(path)-1]
		}
	}
	backtrack(0)
	return results
}

File: daily/Backtracking/90-Subsets-II.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        results = []
        nums.sort()
        def backtrack(start, path):
            results.append(path[:])
            for i in range(start, len(nums)):
                if i > start and nums[i-1] == nums[i]:
                    continue
                path.append(nums[i])
                backtrack(i + 1, path)
                path.pop()
        backtrack(0, [])
        return results